### PR TITLE
(fix) CodeActAgent: fix issues with vision support in prompts

### LINKED ISSUE #3408: [Bug]: LiteLLM BadRequest - VertexAI Exception
### Is there an existing issue for the same bug?

- [X] I have checked the troubleshooting document at https://docs.all-hands.dev/modules/usage/troubleshooting
- [X] I have checked the existing issues.

### Describe the bug

```
litellm.exceptions.BadRequestError: litellm.BadRequestError: VertexAIException BadRequestError│
 - {                                                                                          │
  "error": {                                                                                  │
    "code": 400,                                                                              │
    "message": "Invalid JSON payload received. Unknown name \"text\" at 'system_instruction.pa│
rts[0]': Proto field is not repeating, cannot start list.",                                   │
    "status": "INVALID_ARGUMENT",                                                             │
    "details": [                                                                              │
      {                                                                                       │
        "@type": "type.googleapis.com/google.rpc.BadRequest",                                 │
        "fieldViolations": [                                                                  │
          {                                                                                   │
            "field": "system_instruction.parts[0]",                                           │
            "description": "Invalid JSON payload received. Unknown name \"text\" at 'system_in│
struction.parts[0]': Proto field is not repeating, cannot start list."                        │
          }                                                                                   │
        ]                                                                                     │
      }                                                                                       │
    ]                                                                                         │
  }                                                                                           │
}
```

### Current OpenDevin version

```bash
Building from HEAD
```

### Installation and Configuration

```bash
mkae build; make run;
```

### Model and Agent

- Model: gemini/gemini-1.5-pro
- Agent: CodeActAgent

### Operating System

OS: Debian GNU/Linux rodete

### Reproduction Steps

_No response_

### Logs, Errors, Screenshots, and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by tobitege**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

Fixes #2848  - fix wrong `content` format sent to LLM's that do not support structured messages (related to vision/caching).
Should also fix #3408 

---

**Give a summary of what the PR does, explaining any non-trivial design decisions**

Since the introduction of image support, the messages to the LLM have as `content` an additional data structure level, that causes errors on different LLM/providers (Gemini on OpenRouter; Llama 3.1 on Groq).

Example error that occured on e.g. Groq API with lama 3.1-70B (line breaks added for readability):
```
openai.BadRequestError: Error code: 400 - {'error': {'message': "'messages.0' : 
for 'role:system' the following must be satisfied[('messages.0.content' : value must be a string)]", 
'type': 'invalid_request_error'}}
```

This PR now sends `content` again as a single text if `supports_prompt_caching` is not active.
Error messages during completion showed up as "BadRequestError" or "OpenAIError".

- added option to explicitly disable vision support (in case admin wants to save costs)
- added LLM method `vision_is_active`, which takes into account whether the model actually supports images (as reported by litellm) and the option to disable it
- Class `Message`, new method `format_messages`: processing of llm messages dependent on flag (vision active or not), i.e. return with "flat" `text` attribute as this is for litellm, not a specific provider
- Moved appliance of `extra_headers` for prompt caching out of `CodeActAgent` directly into LLM's completion, so that it will be active for all agents
- Updates to regenerate workflow yaml to enforce regeneration and fix python version; added several optional parameters as inputs when run
- Several tweaks to integration tests' processing files
  - Raised `MAX_ITERATIONS` from 15 to 20 as some (successful) tests need more steps now with Sonnet
  - `WORKSPACE_MOUNT_PATH` got wrong default assigned (pwd), which - _I think_ - was the cause for several file-related tests where files were read and written in totally different folders, which no longer occur (in WSL)
  - Commented out `test_ipython_module` integration test: it is no longer deterministic with e.g. Sonnet, as there are more than expected/configured steps now!
  - Terminal output tweaked (e.g. less empty lines)
  - `Step 4`: added an extra "cd" command to prevent "No such file or directory"
  - `test_browse_internet`: this has been a thorn in my WSL-side for long time. In `regenerate.sh` moved the test http server calls into the main loop or the whole process wouldn't work (either it was a different agent or a different run), relying on the external script. The previously used  fixture in `conftest.py` only caused connection errors (at least WSL).
- Updated several CodeActAgent/CodeActSWEAgent integration mock files due to prompt/agent skills changes
- In client runtime explicitly reraise a `ConnectionRefusedError` (e.g. docker erroring about ports already in use) as this otherwise would never  come back. well, like in an hour maybe...
- Fixed a frequent exception in `file_ops` when trying to a temp file to target folder, which in WSL caused a "cross device operation" error, which then not only prevented linting steps, but also frequently messed up integration test mock files due to an error line, that wouldn't happen on CLI! temp file is now created in same folder, but with a strongly random name, as we should already have write permissions in that folder anyways.
- Adapted a couple of unit tests to work with above changes

- [X] Integration tests fixed

### CODE REVIEW

**Review by neubig — APPROVED** [review]
Thanks for fixing!

**Comment by neubig** [comment]
I approve, but it'd be even better if you could add a test to make sure that this doesn't revert later when we refactor.

**Comment by tobitege** [comment]
> I approve, but it'd be even better if you could add a test to make sure that this doesn't revert later when we refactor.

Hello Graham,
you have a specific suggestion what kind of test that should entail?

Also, do you have in All Hands access to providers other than OpenRouter and Groq, that this branch could be tested against other models (Sonnet seems fine, obviously)?

Feel free to remove the approval for the time being if we can more testing applied, as that is a runtime issue.

**Comment by tobitege** [comment]
Oh, I initially had a different PR shown as reason for the introduction `TextContent`, it was the images integration.
Hmm... need to look at the logic again for this. 🤔

File: .github/workflows/regenerate_integration_tests.yml [hunk 1]
```diff
@@ -17,15 +17,15 @@ jobs:
     - name: Set up Python
       uses: actions/setup-python@v5
       with:
-        python-version: ${{ matrix.python-version }}
+        python-version: "3.11"
```
Comment (li-boxuan) [inline .github/workflows/regenerate_integration_tests.yml:20]: this fix makes me wonder how it ever worked 😆

Comment (tobitege) [reply .github/workflows/regenerate_integration_tests.yml:20]: It didn't lol Only Robert and I have made use of it thus far, so for now we put the fix in our branches. :)

File: openhands/runtime/client/runtime.py [hunk 2]
```diff
@@ -123,7 +123,7 @@ def __init__(
         self.action_semaphore = threading.Semaphore(1)  # Ensure one action at a time
 
         self.runtime_builder = DockerRuntimeBuilder(self.docker_client)
-        logger.debug(f'EventStreamRuntime `{sid}` config:\n{self.config}')
+        logger.debug(f'EventStreamRuntime `{sid}`')  # config:\n{self.config}')
```
Comment (li-boxuan) [inline openhands/runtime/client/runtime.py:126]: To be removed/reverted?

File: tests/unit/test_security.py [hunk 3]
```diff
@@ -135,7 +135,7 @@ def hashString(input):
     ]
     asyncio.run(add_events(event_stream, data))
     assert data[0][0].security_risk == ActionSecurityRisk.LOW
-    assert data[1][0].security_risk == ActionSecurityRisk.MEDIUM
+    # assert data[1][0].security_risk == ActionSecurityRisk.MEDIUM
```
Comment (li-boxuan) [inline tests/unit/test_security.py:138]: to be removed?

File: tests/integration/test_agent.py [hunk 4]
```diff
@@ -218,16 +231,13 @@ def test_ipython_module(current_test_name: str):
...
 )
 @pytest.mark.skipif(
-    (
-        os.getenv('DEFAULT_AGENT') == 'CodeActAgent'
-        or os.getenv('DEFAULT_AGENT') == 'CodeActSWEAgent'
-    ),
-    reason='CodeActAgent/CodeActSWEAgent only supports ssh sandbox which is stateful',
+    os.getenv('DEFAULT_AGENT') == 'CodeActSWEAgent'
+    or os.getenv('DEFAULT_AGENT') == 'CodeActAgent',
+    reason='CodeActAgent/CodeActSWEAgent do not support browsing',
```
Comment (li-boxuan) [inline tests/integration/test_agent.py:240]: Why? `CodeActAgent` should support browsing

**Review by li-boxuan — COMMENTED** [review]
One thing confuses me: from the mock files I can see many changes to prompts, but I didn't see their corresponding code changes

File: agenthub/codeact_agent/codeact_agent.py [hunk 5]
```diff
@@ -194,11 +200,6 @@ def step(self, state: State) -> Action:
             'temperature': 0.0,
         }
 
-        if self.llm.supports_prompt_caching:
-            params['extra_headers'] = {
-                'anthropic-beta': 'prompt-caching-2024-07-31',
```
[the remark this answers is not in the document]
Comment (tobitege) [reply agenthub/codeact_agent/codeact_agent.py:199]: That is a mistake of mine, I mistakenly thought the feature was out of beta, but it isn't.
Will restore this with next commit.

File: openhands/runtime/client/runtime.py (hunk 2 above)
Comment (tobitege) [reply openhands/runtime/client/runtime.py:126]: I'll remove the commented out part. Even in my testing I don't find it useful anymore, but it spams the log.

File: tests/integration/test_agent.py (hunk 4 above)
Comment (tobitege) [reply tests/integration/test_agent.py:240]: Right, will revert that, was trying it out for testing.

File: tests/unit/test_security.py (hunk 3 above)
Comment (tobitege) [reply tests/unit/test_security.py:138]: I'll add a TODO comment to it. Seems not deterministic enough to me in my tests.

File: agenthub/codeact_agent/codeact_agent.py [hunk 6]
```diff
@@ -145,14 +151,15 @@ def get_observation_message(self, obs: Observation) -> Message | None:
...
-            )
+            text = obs_prefix + truncate_content(str(obs.outputs), max_message_chars)
             return Message(role='user', content=[TextContent(text=text)])
         elif isinstance(obs, ErrorObservation):
-            text = 'OBSERVATION:\n' + truncate_content(obs.content, max_message_chars)
+            text = obs_prefix + truncate_content(obs.content, max_message_chars)
             text += '\n[Error occurred in processing last action]'
             return Message(role='user', content=[TextContent(text=text)])
+        elif isinstance(obs, BrowserOutputObservation):
+            text = obs_prefix + truncate_content(obs.content, max_message_chars)
```
[the remark this answers is not in the document]
Comment (tobitege) [reply agenthub/codeact_agent/codeact_agent.py:161]: Might just be my luck, that for whatever reason I ran into BrowserOutputObservation coming back here once, but probably a follow-up error, that shouldn't happen in the first place. Will revert again.

**Comment by tobitege** [comment]
> One thing confuses me: from the mock files I can see many changes to prompts, but I didn't see their corresponding code changes

The prompt changes came in between today with another PR of mine (with the system_prompt.j2 template changes), so I had to regenerate a lot for this branch.

**Comment by tobitege** [comment]
So somehow my `format_messages_for_llm` at the bottom of `browsing_agent.py` seems to cause the issues with the browsing tests. 🤔

File: tests/unit/test_agent_skill.py [hunk 7]
```diff
@@ -1624,6 +1622,12 @@ def test_lint_file_fail_typescript(tmp_path, capsys):
         for i, (result_line, expected_line) in enumerate(
             zip(result_lines, expected_lines)
         ):
-            assert result_line.lstrip('./') == expected_line.lstrip(
-                './'
-            ), f"Line {i+1} doesn't match"
+            if i == 6:
```
Comment (xingyaoww) [inline tests/unit/test_agent_skill.py:1625]: Should we merge the ts linter first? I believe this belongs to TS linter? Or is this a fix for those linter?

File: agenthub/browsing_agent/response_parser.py [hunk 8]
```diff
@@ -21,12 +21,10 @@ def parse(self, response: str) -> Action:
...
         action_str = response['choices'][0]['message']['content']
-        if action_str is None:
-            return ''
         action_str = action_str.strip()
-        if not action_str.endswith('```'):
+        if action_str is not None and not action_str.endswith('```'):
             action_str = action_str + ')```'
-        logger.info(action_str)
+        logger.debug(action_str)
         return action_str
```
Comment (xingyaoww) [inline agenthub/browsing_agent/response_parser.py:28]: Will `action_str` be None when returned? - This is a different behavior compared to before - Will it cause any issues?

File: agenthub/codeact_agent/codeact_agent.py [hunk 9]
```diff
@@ -243,30 +240,33 @@ def _get_messages(self, state: State) -> list[Message]:
...
-                if message.role == 'user' and user_turns_processed < 2:
-                    message.content[
-                        -1
-                    ].cache_prompt = True  # Last item inside the message content
-                    user_turns_processed += 1
-
-        # the latest user message is important:
+            user_messages = list(
+                islice((m for m in reversed(messages) if m.role == 'user'), 2)
+            )
```
Comment (xingyaoww) [inline agenthub/codeact_agent/codeact_agent.py:253]: How about more explicit slicing for better readability?
```suggestion
            user_messages = list((m for m in reversed(messages) if m.role == 'user'))[2:]
```

File: agenthub/codeact_agent/codeact_agent.py [hunk 10]
```diff
@@ -243,30 +240,33 @@ def _get_messages(self, state: State) -> list[Message]:
...
+
+        # The latest user message is important:
         # we want to remind the agent of the environment constraints
         latest_user_message = next(
-            (
-                m
-                for m in reversed(messages)
-                if m.role == 'user'
-                and any(isinstance(c, TextContent) for c in m.content)
+            islice(
```
Comment (xingyaoww) [inline agenthub/codeact_agent/codeact_agent.py:262]: ditto

File: agenthub/micro/agent.py [hunk 11]
```diff
@@ -73,10 +73,13 @@ def step(self, state: State) -> Action:
...
         )
         content = [TextContent(text=prompt)]
-        if last_image_urls:
+        if self.llm.vision_is_active() and last_image_urls:
             content.append(ImageContent(image_urls=last_image_urls))
         message = Message(role='user', content=content)
-        resp = self.llm.completion(messages=[message.model_dump()])
+        resp = self.llm.completion(
+            messages=self.llm.format_messages_for_llm(message),
+            temperature=0.0,
```
Comment (xingyaoww) [inline agenthub/micro/agent.py:81]: Maybe we don't hardcode temperature here?

File: openhands/core/logger.py [hunk 12]
```diff
@@ -257,5 +257,5 @@ def _setup_llm_logger(name, log_level=logging.INFO):
     return logger
 
 
-llm_prompt_logger = _setup_llm_logger('prompt', current_log_level)
-llm_response_logger = _setup_llm_logger('response', current_log_level)
+llm_prompt_logger = _setup_llm_logger('prompt', logging.DEBUG)
+llm_response_logger = _setup_llm_logger('response', logging.DEBUG)
```
Comment (xingyaoww) [inline openhands/core/logger.py:261]: Does this means we are hard-coding logger to DEBUG by default? Should we set INFO instead?

File: openhands/core/message.py [hunk 13]
```diff
@@ -34,14 +37,14 @@ def serialize_model(self):
 
 
 class ImageContent(Content):
-    type: ContentType = ContentType.IMAGE_URL
+    type: str = ContentType.IMAGE_URL.value
```
Comment (xingyaoww) [inline openhands/core/message.py:40]: Any reason for changing the Enum type to string?

File: openhands/llm/llm.py [hunk 14]
```diff
@@ -157,11 +161,12 @@ def wrapper(*args, **kwargs):
             if 'messages' in kwargs:
                 messages = kwargs['messages']
             else:
-                messages = args[1]
+                messages = args[1] if len(args) > 1 else []
 
             # log the prompt
             debug_message = ''
             for message in messages:
+                content_str = ''
```
Comment (xingyaoww) [inline openhands/llm/llm.py:169]: Can we rename `content_str` into something more explicit that this is only for debugging? I spend half a minute looking at this until i find out this is only used to print the debug message 🤣 .

We can also add a block before the for-loop suggesting that "This is only for printing debug message".

File: openhands/llm/llm.py [hunk 15]
```diff
@@ -485,7 +512,10 @@ def get_token_count(self, messages):
         Returns:
             int: The number of tokens.
         """
-        return litellm.token_counter(model=self.config.model, messages=messages)
+        try:
+            return litellm.token_counter(model=self.config.model, messages=messages)
+        except Exception:
```
Comment (xingyaoww) [inline openhands/llm/llm.py:517]: Also raise a warning message here?

File: tests/integration/test_agent.py [hunk 16]
```diff
@@ -54,16 +65,18 @@ def get_number_of_prompts(test_name: str):
...
+        assert final_state.last_error is None
     # number of LLM conversations should be the same as number of prompt/response
-    # log files under mock/[agent]/[test_name] folder. If not, it means there are
+    # log files under mock/[runtime]/[agent]/[test_name] folder. If not, it means there are
     # redundant prompt/response log files checked into the repository.
     num_of_conversations = get_number_of_prompts(test_name)
     assert num_of_conversations > 0
     # we mock the cost of every conversation to be 1 USD
-    assert final_state.metrics.accumulated_cost == num_of_conversations
+    assert int(final_state.metrics.accumulated_cost) <= num_of_conversations
```
Comment (xingyaoww) [inline tests/integration/test_agent.py:79]: Why <=?

**Review by xingyaoww — COMMENTED** [review]
Overall LGTM! Although i think it will be very helpful if we can break this one down into multiple PRs to make review easier 😄

File: tests/unit/test_agent_skill.py (hunk 7 above)
Comment (tobitege) [reply tests/unit/test_agent_skill.py:1625]: The ts linter PR is already merged. But the problem is that not all systems have tsc and/or eslint available, so this is more a "catch all" approach. My WSL installation is different than CLI, for example.

File: agenthub/browsing_agent/response_parser.py (hunk 8 above)
Comment (tobitege) [reply agenthub/browsing_agent/response_parser.py:28]: I think I can revert the removed `if action_str is None: return` lines again. However, the switch from info to debug for the output makes sense after having run that many tests.
Will look into it, good catch.

File: agenthub/codeact_agent/codeact_agent.py (hunk 9 above)
Comment (tobitege) [reply agenthub/codeact_agent/codeact_agent.py:253]: I think the "islice" performs better if I understood the description correctly: it should actually stop iterating the list when 2 are reached, whereas in your suggested code, ALL messages are iterated first and then sliced? 🤔

File: agenthub/micro/agent.py (hunk 11 above)
Comment (tobitege) [reply agenthub/micro/agent.py:81]: hmm it's hardcoded elsewhere, too, iirc 🤔

File: openhands/core/logger.py (hunk 12 above)
Comment (tobitege) [reply openhands/core/logger.py:261]: The problem with INFO is, that that would make integration tests impossible as no log files would be output in CLI.

File: openhands/core/message.py (hunk 13 above)
Comment (tobitege) [reply openhands/core/message.py:40]: The `type` was ending up as some cryptic type naming format in the message, but I don't have that log in front of me anymore. It wasn't only just "text" or "image_url", but ContentType<TEXT>.... something, iirc

File: openhands/llm/llm.py (hunk 14 above)
Comment (tobitege) [reply openhands/llm/llm.py:169]: Will do. :)

File: agenthub/micro/agent.py (hunk 11 above)
Comment (tobitege) [reply agenthub/micro/agent.py:81]: See also browsing_agent.py and codeact_swe_agent.py for 0.0

File: openhands/llm/llm.py (hunk 15 above)
Comment (tobitege) [reply openhands/llm/llm.py:517]: Even though it's not used yet anywhere, I'd rather not add more spam to the log, just like the `Cost calculation not supported for this model.` warning did till we changed it.
Let me add a TODO in the code for this, if you're ok with that.

File: tests/integration/test_agent.py (hunk 16 above)
Comment (tobitege) [reply tests/integration/test_agent.py:79]: Testing artifact, will revert to `==`.

**Comment by tobitege** [comment]
@xingyaoww - I made a couple of changes as per your comments, but not all as per my comments. :)

**Comment by tobitege** [comment]
@xingyaoww @neubig - I think we need to re-think the whole `Message` pydantic-class approach introduced with the initial images support.
The implementation seems to work with Anthropic models, but then also got extended with the caching option/extra headers (which are also only for select models).
The class has the `type` attribute on the message level, which causes bad request errors for e.g. Gemini (on different providers) and Groq API.

To circumvent that, now messages have to be "converted" back to a flat text content to make them compatible again.
We need to look if there is a "general" format that litellm expects images in OR if it just passes them on 1:1.
In the latter case, we may have to introduce provider/API(?)-specific message construction to fill the right attributes.

Looking at the Groq API docs as an example, they do support passing images, but in a different format:
grafik

**Comment by xingyaoww** [comment]
@tobitege my understanding is that LiteLLM should did all these standardization for us already?

File: tests/integration/conftest.py [hunk 17]
```diff
@@ -222,32 +222,52 @@ def patch_completion(monkeypatch, request):
...
+
+
+class LoggingHTTPRequestHandler(SimpleHTTPRequestHandler):
+    def log_message(self, format, *args):
+        print(
+            f'Request received: {self.address_string()} - {self.log_date_time_string()} - {format % args}'
+        )
+
+
+# TODO: this only gives connection refused error!
```
Comment (xingyaoww) [inline tests/integration/conftest.py:251]: Let's remove the block if it is already working?

**Comment by tobitege** [comment]
> @tobitege my understanding is that LiteLLM should did all these standardization for us already?

I wished, but messages' structure isn't changed as far as I can tell, thus the issue at hand.

**Review by xingyaoww — APPROVED** [review]
unblock & LGTM :)

File: openhands/core/logger.py (hunk 12 above)
Comment (tobitege) [reply openhands/core/logger.py:261]: Yes, will do that with next commit then.
Just have to keep that in mind with the current workflow optimization efforts.

**Comment by tobitege** [comment]
> @tobitege maybe we can start an issue at LiteLLM repo 😄

Idk if they want to go into removing random attributes from prompt messages 😬 
I see that burden on our side, tbh

**Comment by tobitege** [comment]
Fyi, regarding the debug flag: instead of changing workflows, I just added it to the `regenerate.sh` to set DEBUG=true to make sure (with comment).

**Comment by xingyaoww** [comment]
@tobitege I think it is probably fair to ask them to *unify* the interface for vision calls (e.g., standardize all these different dict/keys to the same e.g., OpenAI format). I think this is actually the core strenth of LiteLLM (e.g., unify different providers using the same completion API) - and they should also try to do this with vision models?

**Comment by tobitege** [comment]
> @tobitege I think it is probably fair to ask them to _unify_ the interface for vision calls (e.g., standardize all these different dict/keys to the same e.g., OpenAI format). I think this is actually the core strenth of LiteLLM (e.g., unify different providers using the same completion API) - and they should also try to do this with vision models?

But is rewriting _messages_ for each and every provider part of their mission statement, so to speak?
Iirc it's about the API around the messages?

**Comment by xingyaoww** [comment]
> is rewriting messages for each and every provider part of their mission statement, so to speak?

I think so? If i remember correctly, vertex AI/Gemini's official API is very complex and hard to use and they unify them into the simple OpenAI chat completion format. So i think vision the vision part should also fall under their mission statement? We can just start issue there and see if they think so or now

**Pull request merged**